src/tools/storage-perf/StorageIntegrityTool.cpp

```diff
@@ -0,0 +1,263 @@
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "base/Base.h"
+#include "time/Duration.h"
+#include "storage/client/StorageClient.h"
+#include "dataman/RowReader.h"
+#include "dataman/RowWriter.h"
+
+DEFINE_string(meta_server_addrs, "", "meta server address");
+DEFINE_int32(io_threads, 10, "client io threads");
+
+DEFINE_int32(partition_num, 1024, "partititon for space");
+DEFINE_string(space_name, "test_space", "the space name");
+DEFINE_string(tag_name, "test_tag", "the tag name");
+DEFINE_string(prop_name, "test_prop", "the property name");
+
+DEFINE_int64(first_vertex_id, 1, "The smallest vertex id");
+DEFINE_uint64(width, 100, "width of matrix");
+DEFINE_uint64(height, 1000, "height of matrix");
+
+DECLARE_int32(load_data_interval_secs);
+
+namespace nebula {
+namespace storage {
+
+/**
+ * Integration test based on 'IntegrationTestBigLinkedList' of HBase
+ *
+ * We generate a big circle of data, all node is the vertex, and the vertex have only one
+ * property of the next vertex, so we can validate them by traversing.
+ *
+ * There are some gflags we need to pay attention:
+ * 1. The space's replica must be 1, because we don't have retry in StorageClient, we will
+ *    update it after we suppport preheat. The tag must have only one int property,
+ *    which is prop_name. 
+ * 2. If the space and tag doesn't exists, it will try to create one, maybe you need to set
+ *    load_data_interval_secs to make sure the storage service has load meta.
+ * 3. The width and height is the size of the big linked list, you can refer to the graph below.
+ *    As expected, we can traverse the big linked list after width * height steps starting 
+ *    from any node in the list.
+ */
+class IntegrityTest {
+public:
+    IntegrityTest()
+        : propName_(FLAGS_prop_name)
+        , width_{FLAGS_width}
+        , height_{FLAGS_height}
+        , firstVertexId_{FLAGS_first_vertex_id} {}
+
+    int run() {
+        if (!init()) {
+            return EXIT_FAILURE;
+        }
+        prepareData();
+        if (!validate(firstVertexId_, width_ * height_)) {
+            LOG(INFO) << "Integrity test failed";
+            return EXIT_FAILURE;
+        }
+        LOG(INFO) << "Integrity test passed";
+        return 0;
+    }
+
+private:
+    bool init() {
+        FLAGS_load_data_interval_secs = 10;
+        auto metaAddrsRet = nebula::network::NetworkUtils::toHosts(FLAGS_meta_server_addrs);
+        if (!metaAddrsRet.ok() || metaAddrsRet.value().empty()) {
+            LOG(ERROR) << "Can't get metaServer address, status: " << metaAddrsRet.status()
+                       << ", FLAGS_meta_server_addrs: " << FLAGS_meta_server_addrs;
+            return false;
+        }
+        threadPool_ = std::make_shared<folly::IOThreadPoolExecutor>(FLAGS_io_threads);
+        mClient_ = std::make_unique<meta::MetaClient>(threadPool_, metaAddrsRet.value());
+        CHECK(mClient_->waitForMetadReady());
```

Comment:
return false is better
if (!mClient_->waitForMetadReady()) {
print some log
return false;
}

Reply:
Good suggestion.